llvm/lib/CAS/LazyMappedFileRegion.cpp

```diff
@@ -245,6 +245,23 @@ Expected<LazyMappedFileRegion> LazyMappedFileRegion::create(
   sys::fs::file_status Status;
   if (std::error_code EC = sys::fs::status(File, Status))
     return errorCodeToError(EC);
+
+  // Get different kind of lock depending on whether the file is empty or not:
+  // * If empty, lock the file exclusively so only one process will do the
+  //   initialization.
+  // * If non-emty, use shared/reader locking in case another process is in the
```

Comment:
Do you need to a read lock for this?

I think you can always use a write lock but release early if the file size is larger than 0 when the lock returned.

Reply:
I was thinking:
* Open file
* Write lock
* Read size
  * If 0, hold the lock and create
  * If not 0, release the lock
  
This might not scale with the number of process try to access the database. Maybe it is good to add a read lock (there is also a `tryFileLock` function probably needs to gain the same function).

Reply:
> Do you need to a read lock for this?

Yes, a non-zero size doesn't mean the file is initialized, the reader lock ensures that.

> I think you can always use a write lock but release early if the file size is larger than 0 when the lock returned.

I think this is a bit worse because it's adding more logic when a shared lock is already sufficient, and the shared lock also avoids any blocking in the common case where the database file exists and you have multiple processes trying to open it.

Reply:
How about this:
* Open file
* Acquire a read lock
* Check size, if >0, map the file and done, if == 0, release the lock for a write lock
* Check size, if >0, some other process win the race, you can just read, if == 0, hold write lock and map the file.

This is case, you just need to do one stats for the common case, instead of 2.

Reply:
Good idea! See updated commit.